chrome/content/zotero/xpcom/itemPaneManager.js

```diff
@@ -0,0 +1,335 @@
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2024 Corporation for Digital Scholarship
+					 Vienna, Virginia, USA
+					 https://digitalscholar.org
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+/**
+ * @typedef SectionIcon
+ * @type {object}
+ * @property {string} icon - Icon URI
+ * @property {string} [darkIcon] - Icon URI in dark mode. If not set, use `icon`
+ * @typedef SectionL10n
+ * @type {object}
+ * @property {string} l10nID - data-l10n-id for localization
+ * @property {string} [l10nArgs] - data-l10n-args for localization
+ * @typedef SectionButton
+ * @type {object}
+ * @property {string} type - Button type, must be valid DOMString and without ","
+ * @property {(options: SectionEventHookArgs) => void} onClick - Button click callback
+ * @typedef SectionBasicHookArgs
+ * @type {object}
+ * @property {string} paneID - Registered pane id
+ * @property {Document} doc - Document of section
+ * @property {HTMLDivElement} body - Section body
+ * @typedef SectionUIHookArgs
+ * @type {object}
+ * @property {() => {item: Zotero.Item, mode: string, inTrash: boolean, tabType: string}} getData - Get section data
+ * @property {(l10nArgs: string) => void} setL10nArgs - Set l10n args for section header
+ * @property {(l10nArgs: string) => void} setEnabled - Set pane enabled state
+ * @property {(summary: string) => void} setSectionSummary - Set pane section summary
+ * @property {(buttonType: string, options: {disabled?: boolean; hidden?: boolean}) => void} setSectionButtonStatus - Set pane section button status
+ * @typedef SectionHookArgs
+ * @type {SectionBasicHookArgs & SectionUIHookArgs}
+ * @typedef  {SectionHookArgs & { refresh: () => Promise<void> }} SectionInitHookArgs
+ * A `refresh` is exposed to plugins to allows plugins to refresh the section when necessary,
+ * e.g. item modify notifier callback. Note that calling `refresh` during initialization
+ * have no effect.
+ * @typedef  {SectionHookArgs & { incomingData: { type: "item" | "mode" | "inTrash" | "tabType", value: any } }} SectionDataChangeHookArgs
```

Comment:
This is a pretty weird definition — it took me a while even to figure what it meant.

Do these things change independently? If the `tabType` changes, isn't the item probably also going to change? It seems like, if these are actually independent, they should be separate events (e.g., `onItemChange`), but if not, the handler should just get the new complete state of the pane, with all the necessary properties.

Pulling out `inTrash` as a separate state also seems too specific to me. In line with our other change of passing in `collectionTreeRow`, shouldn't the pane just get notified when there's a new `collectionTreeRow`, and it can check if it's a trash or duplicate or whatever else?

Also, what are the possible values of `mode`? Is that just `edit`/`view`? Something else? If just `edit`/`view`, that might be implied by `collectionTreeRow.editable`?

Seems like this could potentially just always include `tabType`, `collectionTreeRow`, and `item` as properties?

Reply:
Or I guess there isn't always a `collectionTreeRow`, like in the reader, but then maybe `collectionTreeRow` (which could be null) and `editable`? Not that we're implementing things with this API at this time (I think?), but wasn't the point of that other change that we're checking various properties on the tree row, and so plugins sections might have reasons to do the same?

Reply:
> Do these things change independently?

Yes. The data are set one by one by the `ItemDetails`.

> Seems like this could potentially just always include tabType, collectionTreeRow, and item as properties?

It's like *a certain property is to be updated. do you want to do something now?*

Passing the changed variable implicitly indicates what is changed. Like, we probably want to update something when `item` is changed, but other properties are not something we care about.

>  but wasn't the point of that other change that we're checking various properties on collectionTreeRow, and so plugin sections might have reasons to do the same?

Well, previously in these `box` things we do not access collectionTreeRow and use these `inTrash` or something that's set by the container.

What I have in mind with these properties:
1. Replace `mode` with `editable`
2. Remove `inTrash`, which is only used by our `attachmentsBox`. We can simply check if the `item` is in trash, right? So no need to rely on the collection tree.
3. Exclude `tabType` from this `onDataChange`, since it's never set a different value.

Do you think these changes make sense?

Reply:
So then there would just be `item` and `editable`? That sounds good. But we have a counterexample in Libraries & Collections, which needs to know the currently selected library and collection, right? How is that getting notified of the selection change?

Reply:
The `Libraries & Collections` currently use `ZoteroPane.collectionsView`, which is OK I think?

Actually, maybe we should skip the `isCurrent` check (make the collection row unclickable if it is the currently selected collection) in non-library (readers), but it's a separate issue.

Reply:
Oh wait... I think `mode` can be `edit`/`view`/`merge`/`mergeedit`. So we actually cannot replace `mode` with `editable`

Reply:
> The Libraries & Collections currently use ZoteroPane.collectionsView, which is OK I think?

Oh right, I forgot that any time the user is switching between collections the item pane is getting re-rendered anyway. So sure, plugins can always check ZoteroPane.collectionsView or anything else in the window, and we'd only need to pass things that could change without an item re-selection.

> I think mode can be edit/view/merge/mergeedit. So we actually cannot replace mode with editable

Well, `merge` and `mergeedit` are only for a few of our own panes in the conflict-resolution dialog. I'm not sure it makes sense to have a more complicated API for plugins just because of that…

Reply:
If we did only have `item` and `editable`, would it still make sense for those to be passed separately? I'm not sure it's even possible for `editable` to change without an item reselection. So then if it's only `item` changing, or an initial `item` being set, I feel like we could just pass these together.

Reply:
> I'm not sure it's even possible for editable to change without an item reselection.

No, I think. I only know the item `editable` changes when it is trashed. But anyway, plugins can listen to item modify notifier, and thus we don't need this `editable` callback. So, let's provide an `onItemChange`.

Reply:
> I only know the item `editable` changes when it is trashed.

No, that's not right — `item.isEditable()` will return `true` for an item in the trash. The item's editable state isn't the same as the current view's editable state. So I think we do need to provide `editable` one way or another — otherwise many plugins would need to check `collectionTreeRow.editable`.

I was just suggesting that we provide both `item` and `editable` in the same event, since `editable` won't be changing on its own.

Reply:
OK, get it.

Reply:
In that case, maybe `editable` can be from `props.getData()`, instead of together with `item` in a `onDataChange`/`onItemChange` callback?

When the item changes/when rendering, plugins can get the `editable` value and do stuff accordingly.

Reply:
Is there a point to `onItemChange`? Can `onRender` be called multiple times on the same item? Or should plugins just check the item when rendering?

Reply:
> Can onRender be called multiple times on the same item

We have a mechanism to avoid duplicate `onRender` calls - only the first one works.

>  Is there a point to onItemChange? Or should plugins just check the item when rendering?

It's mainly for calling `setEnabled` before rendering, to indicate the section is not enabled and hide the corresponding icon in sidenav. For example, when the item is an attachment, the `info` section is not enabled. It would be too late to do this when it's about to render

Reply:
Ah, OK. Makes sense.

So then just getting `editable` when rendering makes sense. But while we're on the subject, is there a reason for having `getData()`, as opposed to just including those things in the passed properties?

Reply:
> as opposed to just including those things in the passed properties?

I intend to indicate they are read-only properties (don't try to change them), and also dynamic properties (don't try to persist them). Do you think the reason is strong enough or there's a better solution?

Reply:
Well, read-only in what way? `item` is the item — there's nothing really read-only about it. And I think passing them in every event handler makes clear that you should check them each time and not persist them.

But in any case, I don't think `getData()` particularly communicates those things. I'd just include them as properties.

Reply:
> read-only in what way?

Like don't try to `setData` to overwrite `editable` or `item`?

> But in any case, I don't think getData() particularly communicates those things. I'd just include them as properties.

OK, then let's do it as properties.

Reply:
Oh, well we can definitely pass a new object to each handler (`Object.assign({}, obj)`) to make sure one isn't doing something funky.

Reply:
Or I guess get fancy with freezing or whatever

Reply:
Just calling `Object.freeze()` on the props object before passing to the handlers is probably sufficient?